baselines/fed_debug/fed_debug/dataset.py

```diff
@@ -0,0 +1,225 @@
+"""Handle basic dataset creation.
+
+In case of PyTorch it should return dataloaders for your dataset (for both the clients
+and the server). If you are using a custom dataset class, this module is the place to
+define it. If your dataset requires to be downloaded (and this is not done
+automatically -- e.g. as it is the case for many dataset in TorchVision) and
+partitioned, please include all those functions and logic in the
+`dataset_preparation.py` module. You can use all those functions from functions/methods
+defined here of course.
+"""
+
+import logging
+import random
+
+import torch
+from pytorch_lightning.core.datamodule import LightningDataModule
+from torch.utils.data import DataLoader
+
+from fed_debug.dataset_preparation import (
+    clients_data_distribution,
+    prepare_iid_dataset,
+    prepare_niid_dataset,
+    train_test_transforms_factory,
+)
+
+
+class NoisyDataset(torch.utils.data.Dataset):
+    """Dataset with noisy labels."""
+
+    def __init__(self, dataset, num_classes, noise_rate):
+        possible_noise_rates = [
+            0.1,
+            0.2,
+            0.3,
+            0.4,
+            0.5,
+            0.6,
+            0.7,
+            0.8,
+            0.9,
+            1,
+        ]
+        assert (
+            noise_rate in possible_noise_rates
+        ), f"Noise rate must be in \
+            /{possible_noise_rates} but got {noise_rate}"
+        self.dataset = dataset
+        self.num_classes = num_classes
+        self.class_ids = random.sample(
+            range(num_classes), int(noise_rate * num_classes)
+        )
+
+    def __len__(self):
+        """Return the size of total dataset."""
+        return len(self.dataset)
+
+    def __getitem__(self, idx):
+        """Return the item at the given index."""
+        x, y = self.dataset[idx]
+        if y in self.class_ids:
+            # random.seed(idx)
+            y_hat = random.randint(0, self.num_classes - 1)
+            if y_hat != y:
+                y = y_hat
+            else:
+                y = (y + 1) % self.num_classes
+        return x, y
+
+
+class FedDataModule(LightningDataModule):
```

Comment:
Do we really need a class just to return dataloaders? the `lightning` packages seems to throw quite a few warnings. It's only used in to define this class and to set the seeds.

Reply:
resolved